File: lib/ts/src/internal/binary/DecodeKeys.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { BinaryEncoding } from '../../internal/binary/BinaryEncoding.js';
import { BinaryEncodingMissing } from '../../internal/binary/BinaryEncodingMissing.js';

export function decodeKeys(given: any, binaryEncoder: BinaryEncoding): any {
    if (Array.isArray(given)) {
        const result = new Array(given.length);
        for (let index = 0; index < given.length; index += 1) {
            result[index] = decodeKeys(given[index], binaryEncoder);
        }
        return result;
    }

    if (given instanceof Map) {
        const newMap: { [key: string]: any } = {};
        for (const [key, value] of given.entries()) {
            const finalKey = typeof key === 'string' ? key : decodeIntegerKey(key, binaryEncoder);
            newMap[finalKey] = decodeKeys(value, binaryEncoder);
        }
        return newMap;
    }

    return given;
}

function decodeIntegerKey(key: unknown, binaryEncoder: BinaryEncoding): string {
    try {
        return binaryEncoder.decodeTable[key as number].toString();
    } catch {
        throw new BinaryEncodingMissing(key as object);
    }
}

File: lib/java/src/main/java/io/github/telepact/internal/binary/DecodeKeys.java
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package io.github.telepact.internal.binary;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class DecodeKeys {
    static Object decodeKeys(Object given, BinaryEncoding binaryEncoder) {
        if (given instanceof Map<?, ?> m) {
            final var newMap = new HashMap<String, Object>();

            for (final var e : m.entrySet()) {
                final String key;
                if (e.getKey() instanceof final String s) {
                    key = s;
                } else if (e.getKey() instanceof final Number n) {
                    try {
                        key = binaryEncoder.decodeTable[n.intValue()];
                    } catch (RuntimeException ex) {
                        throw new BinaryEncodingMissing(e.getKey());
                    }
                } else {
                    throw new BinaryEncodingMissing(e.getKey());
                }
                final var encodedValue = decodeKeys(e.getValue(), binaryEncoder);

                newMap.put(key, encodedValue);
            }

            return newMap;
        } else if (given instanceof final List<?> l) {
            return l.stream().map(e -> decodeKeys(e, binaryEncoder)).toList();
        } else {
            return given;
        }
    }
}
